### ISSUE TITLE
Modifying errors before they are displayed

### ISSUE THREAD

**Issue by royibernthal** [restify/node-restify#1257]
Hi,

I'd like to be able to change how errors are displayed. At the moment they look like that:
`{"code":"InvalidSession","message":"Invalid Session"}`

I'd like to be able to display something like that instead:
`{"error":{"code":"InvalidSession","message":"Invalid Session"}}`

I created custom errors that extend restify.RestError, which I use in my routes (GET/POST/...) via:
`next( new InvalidSessionError() );`

Is it possible to get my hands on the javascript object before it is stringified to JSON and simply modify it?
Is there some sort of parameter that contains the output? Ideally it'd look like that:
`response.output = { error: response.output };`

I had 2 ideas for how this could be achieved:

1) Custom Formatter - It'd be best if I could do it without having to create a formatter from scratch for application/json, i.e to create a formatter which only modifies the javascript object and then passes it to the existing application/json formatter. I tried my hand at it and failed.

2) Listening to an event via server.on where hopefully I'll be able to modify it somehow. When I listen to 'InternalServer' it is never called.

**Comment by DonutEspresso** [comment]
If you're on 4.x, a custom formatter is probably the easiest way to do it. FWIW, the built in JSON formatter is basically just `JSON.parse()` with a custom output for errors, which is what you're looking to customize anyway. 

For errors, you can listen to `server.on('InvalidSession', ...` for your specific error. If you return other errors like `next(new MyCustomError())` you can listen to them by using the error name, `server.on('MyCustom', ...`

If you're on the 5.x beta, you also have the option to hook into the [server.on('restifyError'...](https://github.com/restify/node-restify/blob/5.x/4TO5GUIDE.md#serveronrestifyerror-) event, which will give you the error object for any type of error that was passed to `next()`. You can then modify the response as you like.

**Comment by royibernthal** [comment]
Thanks for the detailed response.

I can switch to 5.x beta, once I listen to the error, how can I modify the response?

Regarding formatter, is there any example that does something similar? From what I saw the formatter code is more complex and I'm not sure what to edit and where.

**Comment by royibernthal** [comment]
Aren't there any examples out there for such a thing?

**Comment by DonutEspresso** [comment]
Sorry for the delayed reply - it depends on the formatter you are using for that response. i.e., if you are depending the default JSON formatter, all it does is call JSON.stringify(). To change that output, you could simply implement [toJSON](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON()_behavior) for the error you want to format:

```js
server.on('InvalidSession', function(req, res, err, cb) {
  err.toJSON = function() { 
    // implement custom logic to format your error object
  };
  return cb();
});
```

**Comment by royibernthal** [comment]
I tried to override err.toJSON:

```
server.on('InvalidSession', function (req, res, err, cb) {
    console.log('InvalidSession', err.toJSON);
    
    err.toJSON = function() {
        return 'some test string';
    }

    return cb();
});
```

The response of the server remains the same and is not modified to 'some test string'.
console.log is called.

err.toJSON returns undefined in the console log before I set it, so I suppose I'm overriding nothing and that's why it's not working?

In my routes, I'm always sending javascript objects to res.send(), which are automatically stringified to JSON by restify.

Does err.toJSON exist by default or is it dependent on adding a handler like jsonp?

Also, how do I switch to the beta version so that I can listen to "restifyError"? Do I need to set restify version to 5 and npm update or something of the sort?

**Comment by DonutEspresso** [comment]
Hello, it depends on what errors you are sending. The snippet above is valid only for cases where you call `res.send(err)` and err is an instance of

**Comment by royibernthal** [comment]
Hey, that's exactly my case. All my errors inherit restify.RestError, and I call res.send(err).

**Comment by DonutEspresso** [comment]
Your errors are inheriting fine, it's that the base RestError prototype has a predefined `toJSON()` which you are trying to override. Without the override, it falls back on `RestError.prototype.toJSON` (or whatever error.prototype.toJSON) property. This is just your standard JS inheritance. 

There's a migration guide here which we will publish in more detail and more fully when 5.x finally goes live:
https://github.com/restify/node-restify/blob/5.x/4TO5GUIDE.md

But primarily what is breaking is that most of the pieces have been broken out into separate modules and repos (clients, plugins, and errors). CORS is a bit of a beast, and since the implementation was not super great in 4.x, we're looking to drop it from core, instead directing folks towards  [restify-cors-middleware](https://github.com/TabDigital/restify-cors-middleware). 

Unfortunately, I don't use TypeScript, so not sure I can help you with the typings.

**Comment by royibernthal** [comment]
I understand. I'm curious why it didn't work, I'll resume trying once I migrate successfully to beta (it didn't go smooth for me unfortunately)

Breaking restify into module looks like a great idea.
I downloaded the required modules and made the required changes in my project.

I encountered 2 errors:

1) When I try to pass an error either to restify.next or restify.send, I'm getting in npm the following error:
"TypeError: Cannot read property 'call' of undefined"

2) restify-cors-middleware doesn't seem to work for me.

```
var cors = corsMiddleware({ credentials: true });

server.pre(cors.preflight);
server.use(cors.actual);
```

> TypeError: Cannot read property 'ALLOW_HEADERS' of undefined
>     at Object.<anonymous> (project_path\node_modules\restify-cors-middleware\
> src\preflight.js:15:41)
>     at Module._compile (module.js:571:32)
>     at Object.Module._extensions..js (module.js:580:10)
>     at Module.load (module.js:488:32)
>     at tryModuleLoad (module.js:447:12)
>     at Function.Module._load (module.js:439:3)
>     at Module.require (module.js:498:17)
>     at require (internal/module.js:20:19)

Regarding TypeScript - is a member of your team writing the TypeScript typings or is it fan work?

**Comment by DonutEspresso** [comment]
`send()` is a method on the response, and `next()` is only provided in the closure of a handler. I don't believe either of these were exposed on the restify main exports previously. 

I'd check out the cors middleware repo and ask for help there - I know of several teams using it successfully but I do not use it myself. 

RE: TypeScript, AFAIK that is also community driven - I did not even know about it until you mentioned it. :)

**Comment by royibernthal** [comment]
Yes I'm aware of that, these are the functions I'm using.

restify.Next is a Typescript definition of the Next object, provided in the closure of the handler.
send() is res.send() (res is restify.Response, also provided the closure of the handler), my bad for writing restify.send().

Since TypeScript is a statically typed language, these kind of things appear as compile time errors, so it's highly unlikely they would happen.

In other words, I'm doing what needs to be done where it needs to be done.
With that said, any idea what else might be the reason for the errors thrown in the first point of my previous comment?

As for the cors middleware - I'll try to ask in their repo as you suggest.

TypeScript is developed and maintained by Microsoft, though I guess custom typings for different modules are created by the community :)

**Comment by royibernthal** [comment]
I just noticed an error is thrown whenever I create the error instance, regardless if I pass to to res.send() or next() or neither.

Is there even an advantage to "inheriting" restify.RestError from restify-errors (e.g. some log supporting features), or is it completely safe to simply create my own independent class for error? (still imitating restify-errors' format though {code, message})

**Comment by DonutEspresso** [comment]
Can you share a short snippet of just the creation of the error object, independent of how you are consuming it within restify?

**Comment by royibernthal** [comment]
At one place I was still referring the error class via the restify module (which returns null in beta). Fixing this solved the problem.
Regarding the original subject of this issue - I managed to do what I needed, thanks to your explanation:

```
server.on('restifyError', function (req: restify.Request, res: restify.Response, err: any, cb: any) {
    var json: any = err.toJSON();

    err.toJSON = function() {
        return { error: json };
    }

    return cb();
});
```

Thank you :)

**Comment by retrohacker** [comment]
This seems to be resolved! :heart:
